[Merged by Bors] - feat(linear_algebra/affine_space/combination, analysis/convex/combination): basic lemmas about affine combinations, center of mass, centroid

ocfnash opened the pull request:
---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

Split off from other work to simplify later review.

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

ocfnash committed:
feat(linear_algebra/affine_space/combination,analysis/convex/basic): basic lemmas about affine combinations, center of mass, centroid

kbuzzard reviewed (commented):
Very nice!

YaelDillies commented on src/linear_algebra/affine_space/combination.lean:
```diff
@@ -229,6 +229,30 @@ lemma affine_combination_vsub (w₁ w₂ : ι → k) (p : ι → P) :
+
+/-- Viewing a module as an affine space modelled on itself, affine combinations are just linear
+combinations. -/
+@[simp] lemma affine_combination_module
```
In my opinion, the name shouldn't mention the typeclass. What about a plain `affine_combination_eq_linear_combination`?

YaelDillies commented on src/analysis/convex/basic.lean:
```diff
@@ -1464,6 +1465,27 @@ lemma finset.center_mass_mem_convex_hull (t : finset ι) {w : ι → ℝ} (hw₀
+    simp only [hs_card, finset.sum_const, nsmul_eq_mul, mul_inv_cancel, ne.def, not_false_iff,
+      finset.centroid_weights_apply, zero_lt_one], },
+  { simp only [imp_self, forall_const, id.def, finset.mem_coe], },
+end
```
Merge master. Those lemmas should now go in `analysis.convex.combination` (and add the import `linear_algebra.affine_space.combination` there. Makes for a neater import tree 😄).

YaelDillies commented on src/linear_algebra/affine_space/combination.lean:
```diff
@@ -229,6 +229,30 @@ lemma affine_combination_vsub (w₁ w₂ : ι → k) (p : ι → P) :
   s.affine_combination p w₁ -ᵥ s.affine_combination p w₂ = s.weighted_vsub p (w₁ - w₂) :=
 by rw [←affine_map.linear_map_vsub, affine_combination_linear, vsub_eq_sub]
 
+lemma affine_combination_attach (s : finset P) (w : P → k) :
```
Can't this be generalised to any `f : s → P`? The name suggests this at least.

ocfnash replied:
Hmm, I think it can indeed be generalised as long as we assume `f` is injective.

So I guess the statement would become something like:
```lean
lemma affine_combination_attach (s : finset P) (w : P → k) (f : s → P) (hf : function.injective f) :
  s.attach.affine_combination f (w ∘ f) = (image f univ).affine_combination id w :=
```

I'll see about adding this later this afternoon.

ocfnash replied:
Now done.

YaelDillies reviewed (approved):
This is very much needed glue between `affine_combination`, `finset.center_mass` and `finset.centroid`!

Could you also add this lemma? I feel like it belongs in this PR.
```
lemma affine_combination_eq_center_mass {ι : Type*} {t : finset ι} {p : ι → E} {w : ι → ℝ}
  (hw₂ : ∑ i in t, w i = 1) :
  affine_combination t p w = center_mass t w p :=
begin
  rw affine_combination_eq_weighted_vsub_of_point_vadd_of_sum_eq_one _ w _ hw₂ (0 : E),
  simp only [vsub_eq_sub, add_zero, finset.weighted_vsub_of_point_apply, vadd_eq_add, sub_zero],
  rw center_mass_eq_of_sum_1 _ _ hw₂,
end
```
Note that it suggests swapping the arguments of `finset.center_mass`.

YaelDillies committed:
Merge remote-tracking branch 'origin' into center_mass_centroid

YaelDillies committed:
rename to affine_combination_eq_linear_combination

YaelDillies committed:
add affine_combination_eq_center_mass

YaelDillies committed:
add finset.center_mass_id_mem_convex_hull

YaelDillies committed:
Merge remote-tracking branch 'origin/master' into center_mass_centroid

YaelDillies committed:
follow the convexity refactor

YaelDillies committed:
Merge remote-tracking branch 'origin/master' into center_mass_centroid

ocfnash committed:
Merge branch 'master' into center_mass_centroid

ocfnash committed:
Generalise `finset.affine_combination_attach` (and rename)

ocfnash committed:
merge master

ocfnash committed:
Clean up after merging master

eric-wieser commented on src/algebra/big_operators/basic.lean:
```diff
@@ -1013,6 +1014,11 @@ calc ∏ a in s, f (g a)
+@[to_additive]
+lemma prod_comp_of_injective [decidable_eq γ] (f : γ → β) (g : α → γ) (hg : function.injective g) :
+  s.prod (f ∘ g) = (s.image g).prod f :=
+by { erw [← s.prod_map ⟨g, hg⟩ f, map_eq_image], congr, }
```
Isn't this a special case of `finset.prod_image`?

ocfnash replied:
Oh, right!

Thanks.

eric-wieser commented on src/data/finset/basic.lean:
```diff
@@ -2041,6 +2041,9 @@ iff.intro
 eq_of_veq $ by rw [image_val, attach_val, multiset.attach_map_val, erase_dup_eq_self]
 
+@[simp] lemma attach_image_coe [decidable_eq α] {s : finset α} : s.attach.image coe = s :=
+finset.attach_image_val
```
Does anything break if we add a simp lemma that says `subtype.val = coe`? If that existed you wouldn't need this lemma.

ocfnash replied:
I'm not sure if anything breaks (because I'd have to wait for a long build) but I don't quite see how `attach_image_coe` follows from:
```lean
@[simp] lemma subtype_val_eq_coe {α : Sort*} {p : α → Prop} :
  subtype.val = (coe : subtype p → α) :=
rfl
```
Are you sure this would suffice?

eric-wieser replied:
You'd also need to mark the lemma above this one `simp`.

ocfnash replied:
Oh I understand now, but things do break if I try placing the lemma in what seemed like the right place.

To be more precise we need a slight reshuffle of what you have suggested: `attach_image_val` would have to be the lemma which follows `by simp` since the LHS of its statement contains `subtype.val`.

In other words, the following would work:
```lean
@[simp] lemma subtype_val_eq_coe {α : Sort*} {p : α → Prop} :
  subtype.val = (coe : subtype p → α) :=
rfl

@[simp] lemma attach_image_coe [decidable_eq α] {s : finset α} : s.attach.image coe = s :=
eq_of_veq $ by rw [← subtype_val_eq_coe, image_val, attach_val, multiset.attach_map_val, erase_dup_eq_self]

lemma attach_image_val [decidable_eq α] {s : finset α} : s.attach.image subtype.val = s :=
by simp only [attach_image_coe, subtype_val_eq_coe]
```

So I tried placing `subtype_val_eq_coe` in core [here](https://github.com/leanprover-community/lean/blob/bef30f3ea338cf0e068d942649a613eabf162920/library/init/coe.lean#L161) however this immediately triggered breakage [here](https://github.com/leanprover-community/lean/blob/bef30f3ea338cf0e068d942649a613eabf162920/library/init/data/fin/ops.lean#L99) (replacing `dsimp` with `dsimp only` fixes it) so I'm wary of fighting for this win.

eric-wieser replied:
I think a safer place would be next to

https://github.com/leanprover-community/mathlib/blob/4b5bf56bd3ca987620453f27491737c8eedc68a7/src/data/subtype.lean#L39

For the sake of this PR, how about you keep your new lemma and mark the `attach_image_val` lemma above as simp too?

ocfnash replied:
I've just pushed https://github.com/leanprover-community/mathlib/pull/9103/commits/e2e4389d46921c4a4283615449d7b8c99b6bb7ec to see if things build with `subtype_val_eq_coe` added where you have suggested.

The build didn't break after 6 minutes (562% CPU) on my machine so there's _some_ change it might work. If this fails, I'll eagerly adopt your compromise.

ocfnash replied:
Aaaaand that failed [here](https://github.com/leanprover-community/mathlib/runs/3730275807#step:6:11).

I've gone with the compromise of just adding `simp` to `attach_image_val`. If https://github.com/leanprover-community/mathlib/pull/9103/commits/3a4d551d06694d70abfef527182a125a62833967 builds successfully, I'll resolve this conversation.

jcommelin replied:
https://leanprover-community.github.io/mathlib_docs/data/subtype.html#subtype.val_eq_coe already exists, right?

ocfnash replied:
> https://leanprover-community.github.io/mathlib_docs/data/subtype.html#subtype.val_eq_coe already exists, right?

Yes. I guess Eric's point is that it should probably state:
```lean
∀ {α : Sort*} {p : α → Prop}, val = (coe : subtype p → α)
```
rather than
```lean
∀ {α : Sort*} {p : α → Prop} {x : subtype p}, x.val = (x : α)
```
as it currently does.

However the change breaks things and I'd need a bit more encouragement to fold this change into this PR.

eric-wieser commented on src/algebra/big_operators/basic.lean:
```diff
@@ -1001,8 +1001,9 @@ finset.strong_induction_on s
-over `b ∈ s.image g` of `f b` to the power of the cardinality of the fibre of `b` -/
-lemma prod_comp [decidable_eq γ] {s : finset α} (f : γ → β) (g : α → γ) :
+over `b ∈ s.image g` of `f b` to the power of the cardinality of the fibre of `b`. See also
+`finset.prod_comp_of_injective`. -/
```
```suggestion
`finset.prod_image`. -/
```

ocfnash committed:
Revert proposed addition `finset.prod_comp_of_injective`

EricW pointed out this is a special case of extant lemma `finset.prod_image`

ocfnash committed:
Try adding `subtype_val_eq_coe`

ocfnash committed:
Revert "Try adding `subtype_val_eq_coe`"

This reverts commit e2e4389d46921c4a4283615449d7b8c99b6bb7ec.

ocfnash committed:
Add `simp` to `finset.attach_image_val`

ocfnash committed:
Merge branch 'master' into center_mass_centroid

ocfnash committed:
Fix up after merging master.

The typeclass `add_comm_group E` was relaxed to `add_comm_monoid E` in ##9356 but
the instance `add_group_is_add_torsor` demands an `add_group` and not just an `add_monoid`.

ocfnash committed:
Upgrade `add_comm_monoid E` to `add_comm_group E` since we have `-1` in the scalars

ocfnash committed:
Merge branch 'master' into center_mass_centroid
